### DISCUSSION TITLE
Loading local soundFont file

### DISCUSSION THREAD

**Discussion by marriusco** [CoderLine/alphaTab#843] [Q&A] [answered]
Am I  doing it right. 

I am trying to open a local sound font file. It looks like is not being applied.
I will step trough the code anyway but I am asking first.

```
   var t = this;
      var a = document.createElement("input");

      a.type = "file", a.accept = ".sf2", a.onchange = function() {
           1 === a.files.length && a.openFile(a.files[0]);
      }, document.body.appendChild(a),
         a.click(),
         document.body.removeChild(a),
         a.openFile=function(e){
            a = new FileReader;
            a.onload=function(e){
                var mob = $("#mobile").length;
                const bytes = new Uint8Array([0]);
                api.settings.player.soundFont = bytes;
//                api.loadSoundFont(bytes); 

                api.updateSettings();
                api.render();
            };
            a.readAsArrayBuffer(e);
        }
```

**Comment by Danielku15** [comment, marked as answer]
There are quite some risky and problematic things in your code you likely need to clean: 

* Reusing `a` for the link and later for the `FileReader` is risky. 
* `const bytes = new Uint8Array([0]);`  is wrong. You are creating here a new byte array with one byte (value=0). You need to access `.result` of the `FileReader` to get the read array buffer. 
* I think it is also wrong to pass `e` into `readAsArrayBuffer` this function needs the `Blob` or `File` object which you want to read as buffer. I recommend building first an example where you are sure you have a proper `ArrayBuffer` or `Uint8Array` before proceeding with forwarding it to alphaTab. If you have problems there, I recommend StackOverflow to get guidance. 
* According to [the docs](https://alphatab.net/docs/reference/settings/player/soundfont/) the `settings.player.soundFont` needs to be a URL. It will not work. But your commented approach should work then as [`loadSoundFont`](https://alphatab.net/docs/reference/api/loadsoundfont/) supports bytes. It's a bit an inconvenience which I might consider to improve that also ArrayBuffer or Uint8Array can be directly set on the `player.soundFont` settings object.

**Reply by marriusco** [reply]
Hi,
Thank for the prompt response. Yes I noticed that. I used a function I had before to open local tabs,
and I paste it and mess it up. [0].. where it came from ?
I am not a JS guy at all, but...  Is fixed and works fine. Here is the final version for reference.

```
function open_local_sf(api)
{
      var t = this;
      var a = document.createElement("input");

      a.type = "file", a.accept = ".sf2", a.onchange = function() {
           1 === a.files.length && a.openFile(a.files[0]);
      }, document.body.appendChild(a),
         a.click(),
         document.body.removeChild(a),
         a.openFile=function(e){
            a = new FileReader;
            a.onload=function(e){
                const bytes = new Uint8Array(e.target.result);
                api.loadSoundFont(bytes);
                api.render();
            };
            a.readAsArrayBuffer(e);
        }
}


```
